packages/block-library/src/navigation/view-modal.js

```diff
@@ -33,4 +33,36 @@ window.addEventListener( 'load', () => {
 		onClose: navigationToggleModal,
 		openClass: 'is-menu-open',
 	} );
+
+	// Close modal automatically on clicking anchor links inside modal.
+	const navigationLinks = document.querySelectorAll(
+		'.wp-block-navigation-item__content'
+	);
+
+	navigationLinks.forEach( function ( link ) {
+		// Ignore non-anchor links and anchor links which open on a new tab.
+		if (
+			! link.getAttribute( 'href' )?.startsWith( '#' ) ||
+			link.attributes?.target === '_blank'
+		) {
+			return;
+		}
+
+		// we need to find the specific parent modal for this link
+		// since .close() won't work without an ID in case we have
+		// mutiple navigation menus in a post/page.
```

Comment:
```suggestion
		// Find the specific parent modal for this link
		// since .close() won't work without an ID if there are
		// multiple navigation menus in a post/page.
```